eip-0008.md

```diff
@@ -0,0 +1,479 @@
+
+#### The Benefits of the JIT costing
+
+1) A miner will never execute more operations for tx validation than it is payed in
+miner's fee outputs, thus block validation costs are strictly limited by sum of fee
+outputs and the miner's validation expenses are fully rewarded, no matter how complex the
+script operations could be. This will enable easy long-term evolution of ErgoTree language
+via soft-forks.
+
+2) When selecting transactions from the pool, a miner can quickly filter out transactions
+that doesn't fit into MaxBlockCost using `feeLimit / costPrice`.
+
+3) In many cases (such as P2PK contracts and others) when the cost of transaction
+validation doesn't depend on the context the cost of `tx` may be exactly predicted by the
+sender. In such cases:
+      - no refund outputs will be created for the transactions spending such "simple"
+      boxes. Thus, the protocol doesn't have overhead in simple cases.
+      - existing applications can be supported (without change) by using reasonable
+      default values when the registers of `feeOut` box are not present.
+
+4) The context independence of scripts and hence the exact prediction of `actCost` is a
+static property of ErgoTree, it can be determined by analysing ErgoTree for which Context
+properties the script uses. This can be controlled and enforced by the sender. Moreover,
+it is possible to extend this property to a wider range of scripts by introducing a
+_validity interval_ of a transaction (see the `R6` register of the `feeOut` box). This is
+an interval of the blockchain heights during which the transaction can be processed. Thus,
+any scripts which run during validation can assume that the current height is within that
+interval, but do not rely on the precise value of the current height. This will make
+`actCost` exactly predictable even for scripts what have conditions on `HEIGHT` (like
+`HEIGHT > deadline`).
+
+5) A block represents a trace of UTXO state changes. Thus depending on a transaction
+validity, either the transaction signed by the sender represents standard (normal) state
+modifications, or the invalid `tx` is interpreted differently as _charge transaction_ thus
+representing different changes in UTXO. This choice is fully deterministic for all
+transactions and thus all validating nodes have a consensus about state changes.
+
+6) In a long term, JIT costing enables an economic incentive to improve performance of
+block validation including script evaluation. The miners are interested to put as much
+transactions in every block as possible to maximize fee payments.
+Considering the pool is full of transactions, the amount of transactions which can be put
+into a new block is limited by two parameters:
+      - Block validation time is limited by 2 minutes tick interval between blocks (which
+      BTW should be adjustable via voting). The time budget for block assembly and
+      validation itself is the fraction of this time (say 30 seconds), as the block must
```

Comment:
Likely a couple of seconds, see GKL15, SPV mining issues in Bitcoin etc.

Reply:
what is GKL15?

Reply:
https://www.iacr.org/archive/eurocrypt2015/90560288/90560288.pdf , see security assumptions on the round length (in comparison with expected block interval)